### PR TITLE
Bob's Shake

### LINKED ISSUE #55: Add Bob's Shake
It's been a while in the thoughts, and we should have a formal mark of wanting to do this. Currently we use Bob's Second as an external tool for analyzing TA help and grading stats, and the logs zygrader produces are used as input for Bob's Second. We can fold this functionality into zygrader for greater ease of maintenance and usage.

This is a bit of a larger undertaking and will involve a fair deal of external conversation.

---

A historical note, this was originally Bob's Minute, and a second iteration was called Bob's Second (to reference it's status as the second iteration and to keep the time reference). The third iteration will be called Bob's Shake ([because shake is a good unit of time too](https://en.wikipedia.org/wiki/Shake_(unit))).

### PR DESCRIPTION
**Pull request by teiknsmith**
There are still two main things to do, but I would like second eyes on this before doing the final testing, and fixing the locks log will need to happen after the merge, so this is ready enough for a review.

### CODE REVIEW

**Comment by natecraddock** [comment]
Could you merge/rebase master?

File: zygrader/bobs_shake.py [hunk 1]
```diff
@@ -0,0 +1,394 @@
+#TODO: Check this against the old Bob's Second
+#TODO: run my 'history fixer' on the lock log
```
Comment (natecraddock) [inline zygrader/bobs_shake.py]: Space after `#`

File: zygrader/bobs_shake.py [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+
+        uniq_item = event_type + student_name + assignment_name
+
+        # FIXME: ta_netid is a lab name if student has a comman in their name
+        #  (this is somewhat fixed in data/lock.py by using csv to write data,
+        #   but old data might still have commas)
+        #  one possible fix (and what I'm thinking I'll do) is to modify the
+        #   the lock log files outside of zygrader,
+        #   just with some separate script, to keep this area clean
+
```
Comment (natecraddock) [inline zygrader/bobs_shake.py]: The comment `# ` spacing here...

**Review by natecraddock — COMMENTED** [review]
Perhaps explain in the source code what "Shake" means

At a high level this code looks really good! I'll take a closer look tomorrow

**Comment by teiknsmith** [comment]
> Could you merge/rebase master?

yeah, I probably should have done that. it has been done now

> Perhaps explain in the source code what "Shake" means

added some changes to help with that too

File: zygrader/bobs_shake.py [hunk 3]
```diff
@@ -0,0 +1,422 @@
...
+    """
+    window = ui.get_window()
+    wait_popup = ui.layers.WaitPopup("Bob's Shake")
+
+    worker = _StatsWorker()
+    Step = namedtuple('Step', ['interactive', 'msg', 'func'])
+    steps = [
+        Step(True, "start time", lambda: worker.select_start_time()),
+        Step(True, "end time", lambda: worker.select_end_time()),
+        Step(True, "select helpqueue csv",
```
Comment (natecraddock) [inline zygrader/bobs_shake.py]: Probably use title case or at least capitalize the messages here

File: zygrader/bobs_shake.py [hunk 4]
```diff
@@ -0,0 +1,422 @@
...
+            event_type = row[1]
+
+        old_format_shift = 1 if is_old_format else 0
+        student_name = row[2 - old_format_shift]
+        assignment_name = row[3 - old_format_shift]
+        ta_netid = row[4 - old_format_shift]
+
+        uniq_item = event_type + student_name + assignment_name
+
+        # FIXME: ta_netid is a lab name if student has a comman in their name
```
Comment (natecraddock) [inline zygrader/bobs_shake.py]: comman -> comma

File: zygrader/bobs_shake.py [hunk 5]
```diff
@@ -0,0 +1,422 @@
...
+
+    @classmethod
+    def from_native_data(cls, row):
+        time_stamp_str = row[0]
+        # TODO: convert to datetime.datetime.fromisoformat
+        # once we are using python 3.7
+        time_stamp = datetime.datetime.strptime(time_stamp_str,
+                                                "%Y-%m-%dT%H:%M:%S.%f")
+
+        # The old lock format did not have an event_type field
```
Comment (natecraddock) [inline zygrader/bobs_shake.py:100]:
```suggestion
        # The old lock format did not have an event_type field
        # (all locks were for labs)
        is_old_format = len(row) == 5
        if is_old_format:
            row.insert(1, "LAB")

        event_type = row[1]
        student_name = row[2]
        assignment_name = row[3]
        ta_netid = row[4]
        
        ...
        
        lock_type = row[5]
```

I am not sure if keeping `row` intact matters, but this is more readable to me

File: zygrader/bobs_shake.py [hunk 6]
```diff
@@ -0,0 +1,422 @@
...
+            else:
+                unmatched_depth -= 1
+                if unmatched_depth == 0:
+                    new_list.append(event)
+        else:
+            if event.is_begin:
+                unmatched_depth += 1
+                new_list.append(event)
+            else:
+                pass
```
Comment (natecraddock) [inline zygrader/bobs_shake.py]: Remove

File: zygrader/data/__init__.py [hunk 7]
```diff
@@ -122,3 +122,38 @@ def write_class_sections(class_sections):
...
+    SharedData.TAS = tas
+
+    tas_json = []
+
+    for ta in tas:
+        tas_json.append(ta.to_json())
+
+    path = SharedData.get_ta_data()
+    with open(path, "w") as _file:
+        json.dump(tas_json, _file, indent=2)
```
Comment (natecraddock) [inline zygrader/data/__init__.py]: End files with trailing newline

**Review by natecraddock — COMMENTED** [review]
Looks good to me. I'd like to take one last look before you commit (since you implied there are a few more things to finish)

**Comment by natecraddock** [comment]
I should mention, there were a few files that were just whitespace cleanups - that can go in now (no need to keep in this branch and PR)

File: zygrader/bobs_shake.py (hunk 3 above)
Comment (teiknsmith) [reply zygrader/bobs_shake.py]: for the interactive steps, the msg is never presented, so I was purposefully leaving these lower case. they only really serve to help in reading the code, and I'm realizing the func name does this well enough, so I'll actually just make them empty I think

File: zygrader/bobs_shake.py (hunk 4 above)
Comment (teiknsmith) [reply zygrader/bobs_shake.py]: this comment will be deleted in the final steps of finishing this when I csv-ify the lock log file, but fair catch

File: zygrader/bobs_shake.py (hunk 5 above)
Comment (teiknsmith) [reply zygrader/bobs_shake.py:100]: it doesn't matter if row stays intact, and I agree. I'll change this

File: zygrader/bobs_shake.py (hunk 6 above)
Comment (teiknsmith) [reply zygrader/bobs_shake.py]: I had that to make it clear that nothing needs to happen there if the event is not begin. but a comment would probably be even better

**Comment by teiknsmith** [comment]
this is now in its (hopefully) final form, with the exception of one TODO which is just a reminder for after this gets added to an update

> I'd like to take one last look before you commit (since you implied there are a few more things to finish)

so this can be done now

**Review by natecraddock — APPROVED** [review]
Perhaps do a squash commit so the log is a bit more clear and not full of various back-and-forth changes

Looks good!

**Pull request merged**